# Lab Notebook for Bi621
## Isis Diaz


### | August 23 of 2022 |
### Assignment 1. PS7
#### The end goal is to create a blast-based Reciprocal Best Hit (RBH) pipeline, calling all 1-to-1 RBH alignments between human and zebrafish proteins. The last step (calling 1-to-1 RBH).
Collaborators:
-
Programs Used:
Python 3.10.4
Modules Used:
-

To start writing the code, the first thing i need to do is read through the file to remember the general structure and then I can make my strategy. In this following steps I will be using the output generated by Leslie Coonrod, to make sure I'm working with correct data and later I will re-run with the output I generated in PS7.

```
cat /projects/bgmp/shared/Bi623/PS7_RBH_Bi623/H_to_zfishdb.blastp | head -1
```
This will show us the following line

```
ENSP00000488212	ENSDARP00000152670	29.73	111	73	4	5	113	1	108	4e-09	51.6
```
Where the protein is column 1 is the protein ID and column 11 is the E-value which is the values we are interested for this assignment.

So the next thing we could do for our analysis is to only keep the values for protein ID, gene ID and e-value. *Explore later*

Working on first sorting by protein name and then by e-value. I'm using the sort command on the batch so It's easier and clearier before i start working on a python code.

To sort two columns I used the following command
```
cat /projects/bgmp/shared/Bi623/PS7_RBH_Bi623/H_to_zfishdb.blastp | sort -k1,1 -k11n,11 | head -100
```

But this isn't going to work since the e-value is in notation and the command doesn't identified it as a full number (it will only take into account the number before the "e"), so changing the ```n``` for a ```g``` might be more appropiate.

```
cat /projects/bgmp/shared/Bi623/PS7_RBH_Bi623/H_to_zfishdb.blastp | sort -k1,1 -k11g,11 | head -100
```
* Note: ENSP is protein of human and ENSDARP is protein of Zebrafish *
 


### | August 31 of 2022 |
### QAA Part2
#### To do

Working on QAA Part2 step 5, I need to obtain the adapter sequences to perform triming. For this you would use the data obtained from running fastqc, in this case you would look at the data.txt and look for:
```
ctr+F | Overrepresented sequences
```
If your overrepresented sequence test pass you will need to look directly at the Ilumina site and revise each sequence for your data type and do a grep analysis to revise if that is the correct adapter.

```
(QAA) [isisd@talapas-ln1 demultiplexed]$ zcat 7_2E_fox_S6_L008_R1_001.fastq.gz | sed -n "2~4p" | grep "AGATCGGAAGAGCACACGTCTGAACTCCAGTCA" | wc -l
6710
(QAA) [isisd@talapas-ln1 demultiplexed]$ zcat 7_2E_fox_S6_L008_R2_001.fastq.gz | sed -n "2~4p" | grep "AGATCGGAAGAGCACACGTCTGAACTCCAGTCA" | wc -l
0
(QAA) [isisd@talapas-ln1 demultiplexed]$ zcat 7_2E_fox_S6_L008_R1_001.fastq.gz | sed -n "2~4p" | grep "AGATCGGAAGAGCGTCGTGTAGGGAAAGAGTGT" | wc -l
0
(QAA) [isisd@talapas-ln1 demultiplexed]$ zcat 7_2E_fox_S6_L008_R2_001.fastq.gz | sed -n "2~4p" | grep "AGATCGGAAGAGCGTCGTGTAGGGAAAGAGTGT" | wc -l
7207
```

In the example above you can see the example for 7_2E_fox_S6_L008 for R1 and R2 to be certain that the correct adapters in the correct files. Once we have demonstrated that we have the correct adapters we can use cutadapt to remove them from our sequences.

```
cutadapt -a AGATCGGAAGAGCACACGTCTGAACTCCAGTCA  -o 7_2E_fox_S6_L008_R1_001_cutadapt.fastq.gz /projects/bgmp/shared/2017_sequencing/demultiplexed/7_2E_fox_S6_L008_R1_001.fastq.gz 
```

### | September 1 of 2022 |
### QAA Part2

-a doesn't work for paired end adapters!. Luiza informed me after talking with Leslie Coonrod

### | September 5 of 2022 |
### QAA Part2

The cutadapt command that i ran is suposedly wrong, because i forgot to consider that the reads are pairended, while the command above is for 3' end only. 

cutadapt --pair-adapters -a AAAAA -a GGGG -A CCCCC -a TTTT -o out.1.fastq -p out.2.fastq in.1.fastq in.2.fastq

```
cutadapt --pair-adapters -a AGATCGGAAGAGCACACGTCTGAACTCCAGTCA -A AGATCGGAAGAGCGTCGTGTAGGGAAAGAGTGT -o 7_2E_fox_S6_L008_R1_001_cutadapt.fastq.gz -p 7_2E_fox_S6_L008_R2_001_cutadapt.fastq.gz /projects/bgmp/shared/2017_sequencing/demultiplexed/7_2E_fox_S6_L008_R1_001.fastq.gz /projects/bgmp/shared/2017_sequencing/demultiplexed/7_2E_fox_S6_L008_R2_001.fastq.gz 
```
```
cutadapt --pair-adapters -a AGATCGGAAGAGCACACGTCTGAACTCCAGTCA -A AGATCGGAAGAGCGTCGTGTAGGGAAAGAGTGT -o 19_3F_fox_S14_L008_R1_001_cutadapt.fastq.gz -p 19_3F_fox_S14_L008_R2_001_cutadapt.fastq.gz /projects/bgmp/shared/2017_sequencing/demultiplexed/19_3F_fox_S14_L008_R1_001.fastq.gz /projects/bgmp/shared/2017_sequencing/demultiplexed/19_3F_fox_S14_L008_R2_001.fastq.gz
```

After running the program I wanted to revise the word count to see how much proportion of the reads was removed. 

#### Original Files
```
zcat /projects/bgmp/shared/2017_sequencing/demultiplexed/7_2E_fox_S6_L008_R1_001.fastq.gz | wc
21113700 26392125 1422754671
```

```
zcat /projects/bgmp/shared/2017_sequencing/demultiplexed/7_2E_fox_S6_L008_R2_001.fastq.gz | wc
21113700 26392125 1422754671
```

#### Cutadapt Files

```
zcat 7_2E_fox_S6_L008_R1_001_cutadapt.fastq.gz | wc
21113700 26391619 1420736175
```

```
zcat 7_2E_fox_S6_L008_R2_001_cutadapt.fastq.gz | wc
21113700 26391619 1420734765
```

#### Output Cutadapt

=== Summary 7_2E_fox_S6_L008===

Total read pairs processed:          5,278,425
  Read 1 with adapter:                  56,771 (1.1%)
  Read 2 with adapter:                  56,771 (1.1%)
Pairs written (passing filters):     5,278,425 (100.0%)

Total basepairs processed: 1,066,241,850 bp
  Read 1:   533,120,925 bp
  Read 2:   533,120,925 bp
Total written (filtered):  1,064,222,649 bp (99.8%)
  Read 1:   532,111,677 bp
  Read 2:   532,110,972 bp

=== Summary 19_3F_fox_S14_L008===

Total read pairs processed:         16,348,255
  Read 1 with adapter:                 193,497 (1.2%)
  Read 2 with adapter:                 193,497 (1.2%)
Pairs written (passing filters):    16,348,255 (100.0%)

Total basepairs processed: 3,302,347,510 bp
  Read 1: 1,651,173,755 bp
  Read 2: 1,651,173,755 bp
Total written (filtered):  3,296,944,816 bp (99.8%)
  Read 1: 1,648,472,709 bp
  Read 2: 1,648,472,107 bp


After running cutadapt now its time to look at trimmomatic. The version being used for this step is 0.39. Bellow I also added the arguments that are going to be used according with the instructions:

- LEADING: quality of 3
- TRAILING: quality of 3
- SLIDING WINDOW: window size of 5 and required quality of 15
- MINLENGTH: 35 bases

Remove leading low quality or N bases (below quality 3) (LEADING:3)
Remove trailing low quality or N bases (below quality 3) (TRAILING:3)
Scan the read with a 5-base wide sliding window, cutting when the average quality per base drops below 15 (SLIDINGWINDOW:5:15
Drop reads below the 35 bases long (MINLEN:35)


Now that the data has been processed, we can create the tables to compare the data. I obtained the general distribution of each file with the following command (example for one file)

```
zcat 7_2E_fox_S6_L008_R2_paired.fq.gz | sed -n '2~4p' | awk '{print(length)}' | sort | uniq -c | sort -n > 7_2E_fox_S6_R2_distribution.txt 

```

I copy all of the files to my computer so I can work on R with the data to graph. I ended creating a data frame for each read (7_2e and 19_3f) and proceded to graph with geom point. But once i look at the data it was difficult to interpret since the distribution was so disperse, so to correct for that i divided agains the max level of counts to obtain a ration and then apply a log2 to be able to appreciate the distribution better. I am gonna upload both graphs to the assignment.

### QAA Part3
Install all of the software required to run Part3

```
conda activate QAA
conda install star (star-2.7.10a)
conda install numpy (numpy-1.23.2)
conda install pysam (pysam-0.19.1)
conda install matplotlib (matplotlib-3.5.3)
conda install htseq (htseq-2.0.2)
```
Then look for gene data for mouse ensemble 107 and a GTF file (gene model). DNA Reference genome for the Mouse and it's GTF file, but the process to obtain them would be the following DNA reference:
```
https://uswest.ensembl.org/index.html | Downloads | FTP Downloads | DNA (FASTA) | 
Mus_musculus.GRCm39.dna.primary_assembly.fa.gz
```
GTF File:
```
https://uswest.ensembl.org/index.html | Downloads | FTP Downloads | GENE SETS | GTF |
Mus_musculus.GRCm39.107.gtf.gz
```

Now that we have the basic data we can start working with STAR. For STAR we require an output directory which i created with the name of the genome and the star version

```
mkdir Mus_musculus.GRCm39.107_star2.7.10a
```

I copied the program i ran for PS8 to create the database for STAR and updated it with the current data. The first code ran is named "star_data.sh"

I have a question for the second part of the STAR, will i aligned the paired data i obtained on PART2 to this mouse genome? Will have to ask tomorrow on open lab!

### | September 6 of 2022 |
### QAA Part3

Yes, we are using our paired data to perform the second part of STAR alignment.

Ran the program to align to star and then ran the flag sam code from PS8 where the results were the following:

aligned_7_2E_Mus_musculus.GRCm39.107_star2.7.10a.Aligned.out.sam
```
The number of reads that are properly mapped are: 9424344
The number of reads that aren't mapped are: 341296
```

aligned_19_3F_Mus_musculus.GRCm39.107_star2.7.10a.Aligned.out.sam
```
The number of reads that are properly mapped are: 30509928
The number of reads that aren't mapped are: 1289070
```

The next part of this assignment is to run htseq for stranded yes and no for each file. Below there is an example of htseq use

```
htseq-count --stranded=yes --format sam aligned/aligned_19_3F_Mus_musculus.GRCm39.107_star2.7.10a.Aligned.out.sam Mus_musculus.GRCm39.107.gtf > htseq_output_19_3f.txt
```

```
htseq-count --stranded=reverse --format sam aligned/aligned_19_3F_Mus_musculus.GRCm39.107_star2.7.10a.Aligned.out.sam Mus_musculus.GRCm39.107.gtf > htseq_output_reverse_19_3f.txt

```

Then we have to count the ammount of sequences that aligned to genes. I ran the following command the resulting 4 files:

```
cat htseq_output_19_3f.txt | grep "^ENS" | cut -f 2 | awk '{s+=$1}END{print(s)}'
555542

cat htseq_output_reverse_19_3f.txt | grep "^ENS" | cut -f 2 | awk '{s+=$1}END{print(s)}'
12936768

cat htseq_output_7_2e.txt | grep "^ENS" | cut -f 2 | awk '{s+=$1}END{print(s)}'
182741

cat htseq_output_reverse_7_2e.txt | grep "^ENS" | cut -f 2 | awk '{s+=$1}END{print(s)}'
4027416
```

Its very noticable that there are more GENE alignments in the reverse stranded 